sabnzbd/downloader.py

```diff
@@ -465,18 +496,25 @@ def run(self):
                             request_server_info(server)
                         break
 
-                    article = sabnzbd.nzbqueue.NzbQueue.do.get_article(server, self.servers)
+                    if not server.article_queue:
+                        fetch_amount = server.threads * 2
+                        if fetch_amount > 50:
+                            fetch_amount = 50
+                        server.article_queue = sabnzbd.nzbqueue.NzbQueue.do.get_articles(server, self.servers, fetch_amount)
+                        if not server.article_queue:
+                            break
 
-                    if not article:
-                        break
+                    article = server.article_queue.pop(0)
+
+                    last_busy[serverid] = time.time()
 
                     if server.retention and article.nzf.nzo.avg_stamp < time.time() - server.retention:
-                        # Let's get rid of all the articles for this server at once
-                        logging.info('Job %s too old for %s, moving on', article.nzf.nzo.final_name, server.host)
-                        while article:
-                            self.decode(article, None, None)
-                            article = article.nzf.nzo.get_article(server, self.servers)
-                        break
+                        if self.decode(article, None, None) < LIMIT_DECODE_QUEUE * 0.5:
+                            break # Try to avoid delayed mode
+                        continue
```

Comment:
This part is rewritten mostly because triggering delayed mode can slow down the downloading significantly for a long time, but also because the queue makes it unnecessary to use article.nzf.nzo.get_article. Unfortunately it also makes it messier to keep the "Job %s too old" info message. It would require a test to check if the server has decoded any parts of that nzb previously, otherwise it will be logged may times.